### DISCUSSION TITLE
Checking if I have implemented my back test correctly using a custom data frame

### DISCUSSION THREAD

**Discussion by pranavlal** [kernc/backtesting.py#731] [Q&A] [answered]
Hi all,
I have the following strategy.
1. I buy when my machine learning model tells me to do so. The model is based on a number of indicators. The signal from the model as well as the price ohlc data is contained in a data frame.
2. 2. I sell if the closing price is above 3 times atr at the time I bought.
3. If I have an open position for a stock, I do not buy it again.

I have the following implementation.
``
from backtesting import Backtest
from backtesting import Strategy
class testModel(Strategy):
	def init(self):
		super().init()
		
		
	def next(self):
		super().next()
		if (self.data.signal==1):
			self.buy()
		else:
			for trade in self.trades:
				if self.data.Close>trade.entry_price+(3*self.data.df["atr"][trade.entry_bar]):
					self.sell()
		

	bt = Backtest(df, testModel, commission=.0002,cash=20000,margin=0.1,exclusive_orders=True)
	stats = bt.run()
	print(stats)
``

Have I implemented the above strategy  as I have described it  in words?

**Comment by kernc** [comment, marked as answer]
Instead of `self.data.signal==1`, you'd make the condition be: `self.data.signal[-1] == 1 and not self.position`.
Instead of `self.sell()`, you'd call [`self.position.close()`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Position.close).

**Comment by pranavlal** [comment]
Many thanks. That works better and is the answer I needed.
